chatrooms.lua

```diff
@@ -51,35 +65,72 @@ function SaveToFile( sChatMessage, sRoom )
 end
 
 function SendToRoom( tSelfUser, sRoom, sIncoming )
-	for iIterate = 0, tChatRooms[sRoom].iMaxProfile do
-		local tUsers = Core.GetOnlineUsers( iIterate )
-		if tUsers then
-			for iIndex, tRecipient in ipairs( tUsers ) do
-				if tRecipient.sNick:lower() ~= tSelfUser.sNick:lower() then
-					Core.SendToUser( tRecipient, "$To: "..tRecipient.sNick.." From: "..sRoom.." $"..sIncoming.."|" )
-				end
+	local tCurrentHistory = tChatRooms[sRoom].tChatHistory
+	table.insert( tCurrentHistory, os.date(tConfig.sTimeFormat)..sIncoming )
+	if tCurrentHistory[ tConfig.iMaxHistory + 1 ] do
+		table.remove( tCurrentHistory, 1 )
+	end
+	local tUsers = tChatRooms[sRoom].tUsers
+	if tUsers then
+		for iIndex, sNick in ipairs( tUsers ) do
+			if sNick:lower() ~= tSelfUser.sNick:lower() then
+				Core.SendToNick( sNick, "$To: "..tRecipient.sNick.." From: "..sRoom.." $"..sIncoming.."|" )
 			end
 		end
 	end
-	SaveToFile( sIncoming, sRoom )
 	return true
 end
 
 function UserConnected( tUser )
 	Hide( tUser )
 end
-RegConnected = UserConnected
+
+function RegConnected( tUser )
+	local bIsSubscribed = false
+	for sRoom, tRoom in pairs( tChatRooms ) do
+		if tRoom.iMaxProfile >= tUser.iProfile then
+			table.insert( tRoom.tUsers, tUser.sNick )
+			bIsSubscribed = true
+		end
+	end
+	if not bIsSubscribed then 
+		Hide( tUser ) 
+	end
+end
+
+function RegDisconnected( tUser )
+	for sRoom, tRoom in pairs( tChatRooms ) do
+		if tRoom.iMaxProfile >= tUser.iProfile then
+			DeleteNick( tChatRooms[sRoom], tUser.sNick:lower() )
+		end
+	end
+end
+
+OpConnected, OpDisconnected = RegConnected, RegDisconnected
 
 function OnTimer( iID )
 	Hide()
 end
 
-function Hide()
-	for sIndex, tValue in pairs( tChatRooms ) do
-		local sQuitINFO = "$Quit "..sIndex.."|"
-		Core.SendToProfile( -1, sQuitINFO )
-		for iIterate = (tValue.iMaxProfile + 1), (tConfig.iTotalProfiles - 1) do
-			Core.SendToProfile( iIterate, sQuitINFO )
+function Hide( tUser )
+	for sRoom, tRoom in pairs( tChatRooms ) do
+		local sQuitINFO = "$Quit "..sRoom.."|"
+		if tUser then
+			Core.SendPmToUser( tUser, sQuitINFO )
+		else
+			Core.SendToProfile( -1, sQuitINFO )
+			for iIterate = (tRoom.iMaxProfile + 1), (tConfig.iTotalProfiles - 1) do
+				Core.SendToProfile( iIterate, sQuitINFO )
+			end
+		end
+	end
+end
+
+function DeleteNick( tTable, sDeleteNick )
+	for iIndex, sNick in ipairs( tTable ) do
+		if sNick:lower() == sDeleteNick then
```

Comment:
Instead of making the developer pass lowercase string, call `:lower()` here for `sDeleteNick`